packages/datadog-instrumentations/src/mysql.js

```diff
@@ -0,0 +1,65 @@
+'use strict'
+
+const { AsyncResource } = require('async_hooks')
+const {
+  channel,
+  addHook,
+  bind,
+  bindAsyncResource
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+addHook({ name: 'mysql', file: 'lib/Connection.js', versions: ['>=2'] }, Connection => {
+  const startCh = channel('apm:mysql:query:start')
+  const asyncEndCh = channel('apm:mysql:query:async-end')
+  const endCh = channel('apm:mysql:query:end')
+  const errorCh = channel('apm:mysql:query:error')
+
+  shimmer.wrap(Connection.prototype, 'query', query => function () {
+    const asyncResource = new AsyncResource('bound-anonymous-fn')
+    const cb = bindAsyncResource.call(asyncResource, arguments[arguments.length - 1])
+
+    if (!startCh.hasSubscribers || arguments.length === 1 || typeof cb !== 'function') {
+      if (arguments[0]._callback) {
+        const newCb = arguments[0]._callback
+        arguments[0]._callback = bind(function () {
+          newCb.apply(this, arguments)
+        })
+      }
+      return query.apply(this, arguments)
+    }
+
+    const startArgs = Array.from(arguments)
+    startCh.publish(startArgs)
+
+    arguments[arguments.length - 1] = bind(function (error, result) {
+      if (error) {
+        errorCh.publish(error)
+      }
+      asyncEndCh.publish(result)
+
+      return cb.apply(this, arguments)
+    })
+
+    try {
+      return query.apply(this, arguments)
+    } catch (err) {
+      err.stack
+      errorCh.publish(err)
+
+      throw err
+    } finally {
+      endCh.publish(undefined)
+    }
+  })
```

Comment:
nit: I'm wondering if we can make this kind of callback wrapper into a util in `helpers/instrument`

@bengl does this workflow differ in the common case? (re: err.stack / finally in particular)

Reply:
One thing to keep in mind with adding helpers is that they won't be available when we move the instrumentation to the library, so any helper that we feel is needed should probably be added to iitm or Node.

Reply:
@bmeck we don't really have this cowpath yet, but we can definitely consolidate it later, and like @rochdev said, try to make sure it exists somewhere upstream.